### PR TITLE
Apply --format when listing repository archives, fixes #1166

### LINKED ISSUE #1166: Trailing and leading whitespace and control characters are allowed as archive names
It is possible for borg archive names to have trailing and leading whitespace and control characters.

```bash
$ borg create "/path/to/repo::archive       " /files/to/backup
$ borg create "/path/to/repo::   archive" /files/to/backup
$ borg create "/path/to/repo::new
> line" /files/to/backup
$ borg create /path/to/repo::$'\1' /files/to/backup
$ borg list /path/to/repo
archive                              Tue, 2016-06-14 02:10:48
   archive                           Tue, 2016-06-14 02:11:20
new
line                             Tue, 2016-06-14 02:11:23
                                    Tue, 2016-06-14 02:12:30
```

Obviously, this creates several issues, escpecially trailing whitespaces.

Trailing whitespace
--
Due to `borg list`'s format **It is impossible to know whether an archive name has trailing spaces, let alone how many spaces it has**.

For example let's create an archive with a normal name.
```bash
$ borg create /path/to/repo::archive /files/to/backup
```
Then create an archive with 1 space after the name, and another with 3 spaces after the name.
```bash
$ borg create "/path/to/repo::archive " /files/to/backup
$ borg create "/path/to/repo::archive   " /files/to/backup
```
Listing the repository gives an awkward result.
```bash
$ borg list /path/to/repo
archive                              Tue, 2016-06-14 03:20:01
archive                              Tue, 2016-06-14 03:24:06
archive                              Tue, 2016-06-14 03:28:13
```
The following problems occur:
- All archives appeares to have the same name, which is impossible.
- It is impossible to know that there are trailing whitespace characters.
- It is impossible to know how many trailing spaces there are.

Control characters and leading whitespaces
--
Control characters and leading whitespaces may make `borg list` harder to read and harder to be parsed by scripts.

Possible solutions
--
- Reject control characters and leading and trailing whitespaces when creating a new archive.
- Truncate leading and trailing whitespace when creating a new archive.
- Change format of `borg list`, or allow to change it via `--list-format`.

*Note: times in the shell's output are not actual times.*

### PR DESCRIPTION
**Pull request by Abogical**
This would give custom formatting when running `borg list /path/to/repo` with `--format`. The current `{keys}` for the formatting are:
- `{name}`: name of the archive
- `{time}`: time of creation of the archive
- `{id}`: the ID of the archive

This would also fix #1166 as weird names can be displayed with the correct formatting, for example:
```bash
$ borg create "/path/to/repo::archive " /files/to/backup
$ borg create "/path/to/repo::archive    " /files/to/backup
$ borg list /path/to/repo
archive                              Tue, 2016-06-14 03:20:01
archive                              Tue, 2016-06-14 03:24:06
$ borg list /path/to/repo --format="{name}/"
archive /archive    /
```
The slash can now be used a seperator as it is not allowed in archive names. This makes the trialing whitespace visible. This also allows easier parsing of `borg list /path/to/repo`, even with weird names.

### CODE REVIEW

File: src/borg/archiver.py [hunk 1]
```diff
@@ -738,6 +738,14 @@ def do_mount(self, args, repository, manifest, key):
...
     def do_list(self, args, repository, manifest, key):
         """List archive or repository contents"""
+        if not hasattr(sys.stdout, 'buffer'):
+                    # This is a shim for supporting unit tests replacing sys.stdout with e.g. StringIO,
+                    # which doesn't have an underlying buffer (= lower file object).
+            def write(bytestring):
+                        sys.stdout.write(bytestring.decode('utf-8', errors='replace'))
+        else:
+            write = sys.stdout.buffer.write
+
```
Comment (ThomasWaldmann) [inline src/borg/archiver.py]: please do this code block move in a separate commit.

File: src/borg/archiver.py [hunk 2]
```diff
@@ -751,20 +759,19 @@ def do_list(self, args, repository, manifest, key):
...
-                if args.short:
+                if args.format:
+                    archive_info_keys.update({'name': archive_info.name,
+                                              'id': bin_to_hex(archive_info.id),
+                                              'time': format_time(to_localtime(archive_info.ts))})
+                    write(safe_encode(format_line(args.format, archive_info_keys)))
+                elif args.short:
                     print(archive_info.name)
                 else:
                     print(format_archive(archive_info))
```
Comment (ThomasWaldmann) [inline src/borg/archiver.py]: please implement this in a similar way as it was done for the item listing:
- implement a ArchiveFormatter class
- process short and standard format by just using the respective format string

File: src/borg/testsuite/archiver.py [hunk 3]
```diff
@@ -1132,6 +1132,17 @@ def test_list_prefix(self):
...
+        self.cmd('init', self.repository_location)
+        self.cmd('create', self.repository_location + '::test-1', src_dir)
+        self.cmd('create', self.repository_location + '::test-2', src_dir)
+        output_1 = self.cmd('list', self.repository_location)
+        output_2 = self.cmd('list', '--format', '{name:<36} {time} [{id}]{NL}', self.repository_location)
+        self.assertEqual(output_1, output_2)
+        output_1 = self.cmd('list', '--short', self.repository_location)
+        output_2 = self.cmd('list', '--format', '{name}{NL}', self.repository_location)
+        self.assertEqual(output_1, output_2)
+
```
Comment (ThomasWaldmann) [inline src/borg/testsuite/archiver.py]: if you have done the above requested changes (using an ArchiveFormattter), some of these tests get pointless and need to actually check whether the output looks as expected (not just test for equality with some other output generated in the same way).

**Comment by ThomasWaldmann** [comment]
In general: good and straightforward idea, just see my remarks about implementation to stay consistent.

**Comment by Abogical** [comment]
Thanks for the feedback, I'll work on this first.

**Comment by Abogical** [comment]
I modified the `{keys}`:
- `{bname}` is the raw archive name with the weird characters
- `{name}` is the archive name with weird characters removed.

keys_help() for this however is not implemented. I would wait for your feedback before I go further.

**Comment by ThomasWaldmann** [comment]
When thinking about it: we maybe want to use same keys all the time, in different contexts. If there is a context not limited to archives, it might be better to name it `barchive` and `archive`.

`name` is too generic, could be also a filename or any other name.

**Comment by ThomasWaldmann** [comment]
https://travis-ci.org/borgbackup/borg/jobs/138930130 pep8 checker is unhappy.

**Comment by Abogical** [comment]
I made those changes. pep8 should be happy now.

**Comment by ThomasWaldmann** [comment]
LGTM. Merge?

**Comment by Abogical** [comment]
You can merge, just don't forget to document it and implement `keys_help()` for ArchiveFormatter.

**Comment by ThomasWaldmann** [comment]
Ah, right, so it is not complete yet. Please document it and add the missing functionality.

**Comment by Abogical** [comment]
Alright I'll do that.

**Comment by Abogical** [comment]
Can I ask why ItemFormatter.keys_help() doesn't just return a fixed string?

**Comment by ThomasWaldmann** [comment]
good question. ask the author. git blame ...

**Comment by Abogical** [comment]
@enkore, can you tell us why?

**Comment by enkore** [comment]
It's been a while, but looking at the code I'm sure that I wrote it that way to enable subclassing. And probably since I didn't want to repeat the key names / avoid typos between docs and what's actually used.

**Comment by Abogical** [comment]
I've added documentation. It's ready to be merged now.

**Comment by Abogical** [comment]
BTW, was I supposed to make the PR to `1.0-maint` instead of `master`?

**Comment by ThomasWaldmann** [comment]
no, master is fine.
